### ISSUE TITLE
a stack overflow on windows in faulthandler._fatal_error()

### ISSUE THREAD

**Issue by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [python/cpython#75864]
BPO | [31683](https://bugs.python.org/issue31683)
--- | :---
Nosy | @vstinner, @orenmn
PRs | <li>python/cpython#3878</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-10-09.16:06:26.945>
created_at = <Date 2017-10-03.21:40:12.194>
labels = ['extension-modules', '3.7', 'type-crash']
title = 'a stack overflow on windows in faulthandler._fatal_error()'
updated_at = <Date 2017-10-09.16:06:26.943>
user = 'https://github.com/orenmn'
```

bugs.python.org fields:
```python
activity = <Date 2017-10-09.16:06:26.943>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-10-09.16:06:26.945>
closer = 'vstinner'
components = ['Extension Modules']
creation = <Date 2017-10-03.21:40:12.194>
creator = 'Oren Milman'
dependencies = []
files = []
hgrepos = []
issue_num = 31683
keywords = ['patch']
message_count = 4.0
messages = ['303651', '303663', '303711', '303982']
nosy_count = 2.0
nosy_names = ['vstinner', 'Oren Milman']
pr_nums = ['3878']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue31683'
versions = ['Python 3.7']
```

</p></details>

**Comment by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [comment]
On my 64-bit Windows 10, the echo here would print -1073741571:
python -c "import faulthandler; faulthandler.\_fatal_error(b'a' * 2 ** 22)"
echo %errorlevel%

This is code c00000fd, which windbg describes as 'Stack overflow'.
This happens because Py_FatalError() (in [Python/pylifecycle.c](https://github.com/python/cpython/blob/main/Python/pylifecycle.c)) does the following
(on Windows only):
    len = strlen(msg);

    /* Convert the message to wchar_t. This uses a simple one-to-one
    conversion, assuming that the this error message actually uses ASCII
    only. If this ceases to be true, we will have to convert. */
    buffer = alloca( (len+1) * (sizeof *buffer));
    for( i=0; i<=len; ++i)
        buffer[i] = msg[i];


Note that (IIUC) running the aforementioned cmd wouldn't cause a post-mortem
debugger to pop-up, because faulthandler_fatal_error_py()
(in [Modules/faulthandler.c](https://github.com/python/cpython/blob/main/Modules/faulthandler.c)) first calls faulthandler_suppress_crash_report(),
and then calls Py_FatalError().

**Comment by vstinner** [comment]
I proposed [PR 3878](https://github.com/python/cpython/pull/3878) to replace alloca() with a fixed buffer of 256 bytes.

**Comment by vstinner** [comment]
New changeset 8d5a3aad2f805dc0ea40829b751f58aa6c75305d by Victor Stinner in branch 'master':
bpo-31683: Py_FatalError() now supports long error messages (bpo-3878)
https://github.com/python/cpython/commit/8d5a3aad2f805dc0ea40829b751f58aa6c75305d

**Comment by vstinner** [comment]
IMHO this issue is theorical. I don't expect that anyone would call Py_FatalError() with a very long message, so I will not backport the fix to Python 2.7 and 3.6.

Thanks for the bug report Oren Milman!
